veps/sig-compute/decentralized-migration/decentralized-migration-vep.md

```diff
@@ -0,0 +1,308 @@
+
+There is potentially an issue with splitting the migrateState into a `source` and `target` state in the `VirtualMachineInstance`. Any ongoing live migrations are at potential risk.
+
+## Functional Testing Approach
+
+There is no functional difference when doing a split live migration in the same namespace vs doing an existing live migration. So to test the new functionality we need to ensure that existing live migration tests still work, and additional tests are added that explicitly use the source and target flow.
+
+Cross cluster live migration is going to be hard to functionally test in the kubevirtci system since it requires two clusters. But conceptually there is no difference between a live migration between namespaces and live migration between clusters. The only difference is the network connectivity between the virt-handlers. And that detail should be handled outside of KubeVirt and thus we can just test cross namespace live migration to ensure this functionality will work.
+
+Since there will be no functional difference between existing live migration and the new mechanism. All the existing live migration tests will still be relevent. In addition we have to add a few tests that use the new `source` and `target` migration resources in the same namespace, and different namespaces.
+
+## Implementation Phases
+
+The implementation phases will closely follow the proposed state diagrams above.
+- First decentralize the live migration.
+  - Untangle the virt-controller migration controller, to have two flows. Once for the source flow and one for the target flow.
+  - Untangle the virt-handler controller into three separate flows. This will likely be achieved by turning the single controller into three controllers:
+    - regular flow
+    - migration source flow
+    - migration target flow
+  - Keep the same migration resources unmodified. This will ensure there are no regressions during the untangling.
+  - Possibly split the `VirtualMachineInstance` migration state into two states if necessary. If it not necessary this becomes part of the second phase.
+- Second modify the migration resources to have the new fields as proposed. And if not modified already, split the `VirtualMachineInstance` migration state into two states.
+  - Add functional test that mimics a basic migration functional but with the separated migration resources.
+- Third implement the synchronization controller and associated logic. The controller is not need in the first two phases, but once migrating across namespaces some synchronization is required.
+- Fourth Implement the logic for cross namespace migration. This will require adding fields to the migration state to properly handle the flow.
+  - Add a functional test to do a basic migration across namespaces.
+- Fifth To simulate migration across clusters, add a functional test that switches network and storage when migrating between namespaces.
+
+## Feature lifecycle Phases
```

Comment:
Can we really put a feature gate on this? I mean, it will refactor the code so deeply that I'm not sure we can really separate it with a feature gate. Or do we plan to have a code duplication, like 2 files with the 2 version of the code? I know, I'm going into technical details but I really would like to understand how this can be feature gated

Reply:
So the feature gate would be to enable the 'synchronization-controller' and allow the split VirtualMachineInstanceMigration resources. Throughout this entire process we are aiming to keep existing migration flows as they are, and thus that would not be subject to the feature gate.

Reply:
I will throw here the idea, even if it is pretty ugly. If we could have for a while some code duplication, and the feature gate makes the switch between the 2 code versions, and then we gradually integrate the new code into the old one when we are more secure.

I'm well aware that this double(or triple) the reviewer work.
@vladikr wdyt?

Reply:
So regular migrations will not change from a user perspective. Only if we want to do the more elaborate migrations will we see a difference. We can have confidence that migrations still work because of the test suite catching anything that gets broken with the new code.